runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,215 @@
+  ---@private
+  local function pos_to_byte(position)
+    local col = position.character
+    if col > 0 then
+      local line = lines[position.line + 1] or ''
+      local ok, convert_result
+      ok, convert_result = pcall(util._str_byteindex_enc, line, col, client.offset_encoding)
+      if ok then
+        return convert_result
+      end
+      return math.min(#line, col)
+    end
+    return col
+  end
+
+  for _, hint in ipairs(result) do
+    local lnum = hint.position.line
+    hint.position.character = pos_to_byte(hint.position)
+    table.insert(new_hints_by_lnum[lnum], hint)
+  end
+
+  hints_by_client[client_id] = new_hints_by_lnum
+  bufstate.version = ctx.version
+  api.nvim__buf_redraw_range(bufnr, 0, -1)
+end
+
+---@private
+local function resolve_bufnr(bufnr)
+  return bufnr == 0 and api.nvim_get_current_buf() or bufnr
+end
+
+--- Refresh inlay hints for a buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---@param opts (nil|table) Optional arguments
+---  - bufnr (integer, default: 0): Buffer whose hints to refresh
+---  - only_visible (boolean, default: false): Whether to only refresh hints for the visible regions of the buffer
+---
+--- Example:
+--- <pre>vim
+---   autocmd BufEnter,InsertLeave,BufWritePost <buffer> lua vim.lsp.inlay_hint.refresh()
+--- </pre>
+---
+function M.refresh(opts)
+  opts = opts or {}
+  local bufnr = opts.bufnr or 0
+  local only_visible = opts.only_visible or false
+  bufnr = resolve_bufnr(bufnr)
+  M.__explicit_buffers[bufnr] = true
+  local buffer_windows = {}
+  for _, winid in ipairs(api.nvim_list_wins()) do
+    if api.nvim_win_get_buf(winid) == bufnr then
+      table.insert(buffer_windows, winid)
+    end
+  end
+  for _, window in ipairs(buffer_windows) do
+    local first = vim.fn.line('w0', window)
+    local last = vim.fn.line('w$', window)
+    local params = {
+      textDocument = util.make_text_document_params(bufnr),
+      range = {
+        start = { line = first - 1, character = 0 },
+        ['end'] = { line = last, character = 0 },
+      },
+    }
+    vim.lsp.buf_request(bufnr, 'textDocument/inlayHint', params)
+  end
+  if not only_visible then
+    local params = {
+      textDocument = util.make_text_document_params(bufnr),
+      range = {
+        start = { line = 0, character = 0 },
+        ['end'] = { line = api.nvim_buf_line_count(bufnr), character = 0 },
+      },
+    }
+    vim.lsp.buf_request(bufnr, 'textDocument/inlayHint', params)
+  end
+end
+
+--- Clear inlay hints
+---
+---@param client_id integer|nil filter by client_id. All clients if nil
+---@param bufnr integer|nil filter by buffer. All buffers if nil
+function M.clear(client_id, bufnr)
+  local buffers = bufnr and { resolve_bufnr(bufnr) } or vim.tbl_keys(hint_cache_by_buf)
+  for _, iter_bufnr in ipairs(buffers) do
+    M.__explicit_buffers[iter_bufnr] = false
+    local bufstate = hint_cache_by_buf[iter_bufnr]
```

Comment:
We are accumulating lots of module-specific cache logic in the lsp modules. Is there actually a case where the "inlay hints cache" would be cleared but "codelenses cache" would not?

Reply:
I don't think so

Reply:
@justinmk I want codelens to only be updated sometimes and inlay hints to be updated frequently, so I update inlay hints on InsertLeave too, while codelens is only refreshed on BufWritePost

Reply:
> inlay hints to be updated frequently, 
I wonder if you do some change then undo / reundo many times(just for test) how it works. ?  Is it more useful to use `nvim_buf_attach` ` on_line` and `on_changedtick` for more detailed synchronization buffer data for inlay request? (just a thought) recently i use these instead of autocmd event on lspsaga /documentSymbol .works better than autocmd. be careful that callback run very frequently. so use delayed requests to avoid sending multiple requests or resending requests before the end, etc skills can make it better . if now codes works fine nvm

Reply:
auto refresh will be implemented in a follow up PR, this one just has manual refresh - I was talking about autocmds in personal configuration